### DISCUSSION TITLE
Action not visible when canSee = true but canRun = false

### DISCUSSION THREAD

**Discussion by ghost** [laravel/nova-issues#5157] [Q&A] [answered]
- Laravel Version: 9.41.0
- Nova Version: 4.19.3
- PHP Version: 8.1.13

### Description:
I have a destructive action with canSee and canRun rules applied to it. Given that canSee resolves to "true" but canRun to "false", the action is nowhere visible - nor in the details view, neither on the index view.

### Detailed steps to reproduce the issue on a fresh Nova installation:

```php
(new DeactivateResourceAction())
    ->showInline()
    ->canSee(function ($request) {
        if ($request instanceof ActionRequest) {
            return true;
        }
    
        return $this->resource->exists && $this->resource->isActive(); // returns true;
    })
    ->canRun(function ($request, $role) {
        return $role->users_count === 0; // returns false;
    })
```

**Comment by crynobone** [comment]
> Given that canSee resolves to "true" but canRun to "false", the action is nowhere visible - nor in the details view, neither on the index view.

This is expected behavior, almost all selection will check `canSee` and `canRun` before showing available actions (except when you select all matching resources).

**Comment by ghost** [comment]
Thanks @crynobone. 
What's the purpose of canRun then? I got impression that canSee limits the visibility, whereas canRun determines whether the action can be executed or not. I was expecting that action will show up but when clicked, it would show an error message.

**Reply by crynobone** [reply]
1. `canSee` doesn't pass `$model` instance as second parameter, using `$this->resource` has always been hacky solution.
2. Your expectation are based on Laravel Nova 3, where it has been a long requested feature to hide Action when user aren't authorized to run it. And this is now become the current behavior.

**Comment by sanderbaas** [comment]
I don't understand either. According to the documentation it should work as salvisb says.

From https://nova.laravel.com/docs/4.0/actions/registering-actions.html#action-visibility
![afbeelding](https://user-images.githubusercontent.com/4152895/207598941-64c3caf2-982b-47e8-a9f3-faf7b99751da.png)

**Comment by Radiergummi** [comment]
@crynobone If I may reignite this discussion, it would be really neat if there was a way to show an actions exists, but "grey it out" to indicate the user cannot run it at this time. I have had several cases now where users were simply not aware some functionality exists because they haven't encountered it yet and just assumed it wasn't possible in the first place.

**Comment by davidhemphill** [comment, marked as answer]
In the next release, we are pushing a change to Actions that will make it more consistent with the expectations for what `canSee` and `canRun` should do. In short, if you can see an action, but  can't run it, it will be visible but disabled in the UI.
